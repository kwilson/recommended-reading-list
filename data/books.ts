import { parseIsbn, parseString, parseNumber } from '../utils/data-parser';

export interface IBookDataResponseItem {
  id: string;
  cover: string;
  isbn: string;
  title: string;
  subtitle: string;
  author: string;
  published: string;
  publisher?: string;
  pages?: number;
  description?: string;
  website?: string;
}

export interface IBookDataResponse {
  books: IBookDataResponseItem[];
}

export interface IBook {
  id: string;
  title: string;
  author: string;
  cover: string;
}

export interface IBookDetails extends IBook {
  isbn: string | null;
  subtitle: string;
  published: string;
  publisher: string | null;
  pages: number | null;
  description: string | null;
  website: string | null;
}

const endpoint = 'https://hokodo-frontend-interview.netlify.app/data.json';

async function getBooksData(): Promise<IBookDataResponse> {
  const data = await fetch(endpoint);
  const values: IBookDataResponse = await data.json();

  return values;
}

export const getBooks = async (): Promise<IBook[]> => {
  try {
    const data = await getBooksData();
    return data.books.map((input) => ({
      author: input.author,
      cover: input.cover,
      id: input.id,
      title: input.title,
    }));
  } catch (e) {
    return [];
  }
};

export const getBooksByAuthor = async (
  author: string,
  exludeIds: string[] = []
): Promise<IBook[]> => {
  try {
    const data = await getBooksData();
    const predicate = (book: IBook) =>
      book.author === author && !exludeIds.includes(book.id);

    return data.books.filter(predicate).map((input) => ({
      author: input.author,
      cover: input.cover,
      id: input.id,
      title: input.title,
    }));
  } catch (e) {
    return [];
  }
};

export const getBook = async (
  id: string | null
): Promise<IBookDetails | null> => {
  try {
    const { books } = await getBooksData();
    const match = books.find((book) => book.id === id);

    if (match) {
      return {
        author: match.author,
        cover: match.cover,
        description: parseString(match.description),
        id: match.id,
        isbn: parseIsbn(match.isbn),
        pages: parseNumber(match.pages),
        published: match.published,
        publisher: parseString(match.publisher),
        subtitle: match.subtitle,
        title: match.title,
        website: parseString(match.website),
      };
    }

    return null;
  } catch (e) {
    return null;
  }
};
