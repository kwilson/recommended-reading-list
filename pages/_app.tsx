import type { AppProps /*, AppContext */ } from 'next/app';

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function AppTemplate({ Component, pageProps }: AppProps) {
  return (
    <div className="container">
      <style jsx global>{`
        html,
        body {
          padding: 0;
          margin: 0;
          font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto,
            Oxygen, Ubuntu, Cantarell, Fira Sans, Droid Sans, Helvetica Neue,
            sans-serif;
        }

        * {
          box-sizing: border-box;
        }

        :root {
          --border-color: black;
        }

        .container {
          padding: 0 1rem;
          margin: 0 auto;
          max-width: 960px;
        }

        .banner {
          padding: 0.5rem 0;
          border-bottom: solid 1px var(--border-color);
        }
      `}</style>

      <div className="banner">Recommended Reading List</div>
      <Component {...pageProps} />
    </div>
  );
}

export default AppTemplate;
