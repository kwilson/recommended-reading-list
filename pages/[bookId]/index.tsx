import Head from 'next/head';
import Link from 'next/link';
import { FunctionComponent, useEffect, useState } from 'react';
import { useRouter } from 'next/router';

import { BookDetails as BookDetailsCmp } from '../../components/BookDetails';
import { LoaderContainer } from '../../components/LoaderContainer';
import { getBook, IBookDetails } from '../../data/books';

export const BookDetails: FunctionComponent = () => {
  const router = useRouter();

  const [book, setBook] = useState<IBookDetails | null>(null);
  const [isLoading, setIsLoading] = useState(true);

  const bookId = router?.query?.bookId as string;

  useEffect(() => {
    if (bookId) {
      setIsLoading(true);
      getBook(bookId)
        .then(setBook)
        .then(() => setIsLoading(false));
    }
  }, [bookId]);

  if (!isLoading && book === null) {
    return (
      <div data-testid="not-found">
        <Head>
          <title>Not Found</title>
          <link rel="icon" href="/favicon.ico" />
        </Head>

        <h1>Not Found</h1>
        <p>Sorry, the book you are trying to view has not been found.</p>

        <p>
          <Link href="/">Back to book list</Link>
        </p>
      </div>
    );
  }

  return (
    <div>
      <Head>
        <title>
          {book?.title} by {book?.author}
        </title>
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <LoaderContainer loading={isLoading}>
        {book && <BookDetailsCmp {...book} />}
      </LoaderContainer>

      <p>
        <Link href="/">Back to book list</Link>
      </p>
    </div>
  );
};

export default BookDetails;
