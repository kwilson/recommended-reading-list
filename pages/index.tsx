import Head from 'next/head';
import { FunctionComponent, useEffect, useState } from 'react';

import { BookList } from '../components/BookList';
import { LoaderContainer } from '../components/LoaderContainer';

import { getBooks, IBook } from '../data/books';

export const Home: FunctionComponent = () => {
  const [books, setBooks] = useState<IBook[]>([]);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    setIsLoading(true);
    getBooks()
      .then(setBooks)
      .then(() => setIsLoading(false));
  }, []);

  return (
    <div>
      <Head>
        <title>Recommended Reading List</title>
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <h1 className="title" data-testid="title">
        All Books
      </h1>

      <LoaderContainer loading={isLoading}>
        <BookList books={books} />
      </LoaderContainer>
    </div>
  );
};

export default Home;
