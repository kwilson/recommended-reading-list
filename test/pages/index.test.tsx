import { render, screen, waitFor } from '@testing-library/react';

import { getBooks, IBook } from '../../data/books';
import { Home } from '../../pages';

jest.mock('../../data/books');
const mockGetBooks = getBooks as jest.Mock<ReturnType<typeof getBooks>>;

describe(`<${Home.name} />`, () => {
  it('renders as expected when loading', async () => {
    const promise = new Promise<IBook[]>((resolve) => {
      setTimeout(resolve, 1000);
    });
    mockGetBooks.mockReturnValueOnce(promise);

    const { asFragment } = render(<Home />);

    await waitFor(() => screen.getByRole('heading'));

    expect(screen.getByTestId('loader')).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders as expected with no data', async () => {
    mockGetBooks.mockResolvedValueOnce([]);

    const { asFragment } = render(<Home />);

    await waitFor(() => screen.getByRole('heading'));

    expect(screen.getByTestId('empty-book-list')).toBeDefined();
    expect(screen.queryByTestId('loader')).toBe(null);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders as expected with data', async () => {
    mockGetBooks.mockResolvedValueOnce([
      {
        id: 'id',
        title: 'title',
        author: 'author',
        cover: 'cover',
      },
    ]);

    const { asFragment } = render(<Home />);

    await waitFor(() => screen.getByRole('heading'));

    expect(screen.queryByTestId('empty-book-list')).toBe(null);
    expect(screen.queryByTestId('loader')).toBe(null);
    expect(asFragment()).toMatchSnapshot();
  });
});
