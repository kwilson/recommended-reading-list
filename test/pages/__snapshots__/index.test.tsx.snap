// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders as expected when loading 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="title"
      data-testid="title"
    >
      All Books
    </h1>
    <div
      class="container isLoading"
    >
      <div
        class="notification"
        data-testid="loader"
      >
        Loading...
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Home /> renders as expected with data 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="title"
      data-testid="title"
    >
      All Books
    </h1>
    <div
      class="container"
    >
      <div
        class="container"
      >
        <div
          data-testid="book-list-items"
        >
          <div
            class="container"
          >
            <div
              class="details"
            >
              <h2
                class="title"
                data-testid="book-title"
              >
                title
              </h2>
              <div
                class="author"
              >
                by 
                <span
                  data-testid="book-author"
                >
                  author
                </span>
              </div>
              <div
                class="link"
                data-testid="book-link"
              >
                <a
                  href="/id"
                >
                  More Info
                </a>
              </div>
            </div>
            <div
              class="cover"
            >
              <img
                alt="title book cover"
                class="coverImage"
                data-testid="book-image"
                src="cover"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Home /> renders as expected with no data 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="title"
      data-testid="title"
    >
      All Books
    </h1>
    <div
      class="container"
    >
      <div
        class="container"
      >
        <div
          data-testid="empty-book-list"
        >
          <p>
            Sorry, there are no books to display.
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
